### PR TITLE
[8.x] Add a cluster listener to fix missing system index mappings after upgrade

### PR DESCRIPTION
**Pull request by ldematte**
This PR modifies `TransportVersionsFixupListener` to include all of compatibility versions (not only TransportVersion) in the fixup.

`TransportVersionsFixupListener` spots the instances when the master has been upgraded to the most recent code version, along with non-master nodes, but some nodes are missing a "proper" (non-inferred) Transport version.
This PR adds another check to also ensure that we have real (non-empty) system index mapping versions.

To do so, it modifies NodeInfo so it carries all of CompatibilityVersions (TransportVersion + SystemIndexDescriptor.MappingVersions).

This was initially done via a separate fixup listener + ad-hoc transport action, but the 2 listeners "raced" to update ClusterState on the same CompatibilityVersions structure; it just made sense to do it at the same time.

The fixup is very similar to https://github.com/elastic/elasticsearch/pull/110710, which does the same for cluster features; plus, it adds a CI test to cover the bug raised in https://github.com/elastic/elasticsearch/issues/112694

Closes https://github.com/elastic/elasticsearch/issues/112694

### CODE REVIEW

**Comment by ldematte** [comment]
Removing auto-backport as this includes a tv change.
Also, not sure if it's worth it - should we backport it to 8.16? 8.15?

**Comment by thecoop** [comment]
I think we should backport it to 8.16.1, given it fixes some erroneous behaviour users have seen. We can use a patch TV to do this change on 8.16.1.

**Comment by ldematte** [comment]
> We can use a patch TV to do this change on 8.16.1.

Will do that, thanks!

File: server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java [hunk 1]
```diff
@@ -98,26 +103,26 @@ public TransportVersionsFixupListener(
     }
 
     class NodeTransportVersionTask implements ClusterStateTaskListener {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java:105]: Maybe rename the class?

File: test/framework/src/main/java/org/elasticsearch/test/rest/RestTestLegacyFeatures.java [hunk 2]
```diff
@@ -184,7 +191,8 @@ public Map<NodeFeature, Version> getHistoricalFeatures() {
             entry(NEW_DATA_STREAMS_INDEX_NAME_FORMAT, Version.V_7_11_0),
             entry(DISABLE_FIELD_NAMES_FIELD_REMOVED, Version.V_8_0_0),
             entry(ML_NLP_SUPPORTED, Version.V_8_0_0),
-            entry(MAPPINGS_UPGRADE_SERVICE_USES_MAPPINGS_VERSION, Version.V_8_11_0)
+            entry(MAPPINGS_UPGRADE_SERVICE_USES_MAPPINGS_VERSION, Version.V_8_11_0),
+            entry(SYSTEM_INDICES_MAPPING_VERSION_IN_CLUSTER_STATE, Version.V_8_11_0)
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/RestTestLegacyFeatures.java:195]: The test itself uses 8.16?

File: qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/SystemIndexMappingUpgradeIT.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    private static String getNodesAttribute(ObjectPath objectPath, String nodeId, String attribute) {
+        try {
+            return objectPath.evaluate("nodes." + nodeId + "." + attribute);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static List<String> getUpgradedNodesAddresses() throws IOException {
```
Comment (thecoop) [inline qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/SystemIndexMappingUpgradeIT.java:98]: There must be a helper method somewhere that does this already?

**Review by thecoop — APPROVED** [review]
Principal looks good, just a few things to tweak

**Comment by thecoop** [comment]
If this is being backported to 8.16, the backport PR should be in place too, and the backport transport version referenced in this PR

File: server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java (hunk 1 above)
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java:105]: `NodeCompatibilityVersionsTask`?
Should I rename the outer class too, to `CompatibilityVersionsFixupListener`?

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java:105]: Yes, that would make sense

File: server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java [hunk 4]
```diff
@@ -151,20 +168,52 @@ private static Map<String, CompatibilityVersions> getCompatibilityVersions(Clust
     public void clusterChanged(ClusterChangedEvent event) {
         if (event.localNodeMaster() == false) return; // only if we're master
 
+        List<Stream<String>> queries = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java:171]: A set will do :)

File: server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java [hunk 5]
```diff
@@ -151,20 +168,52 @@ private static Map<String, CompatibilityVersions> getCompatibilityVersions(Clust
...
+         * This whole class will be removed in ES v9.
+         */
+        queries.add(
+            event.state()
+                .nodes()
                 .stream()
-                .filter(e -> e.getValue().transportVersion().equals(INFERRED_TRANSPORT_VERSION))
-                .map(Map.Entry::getKey)
-                .collect(Collectors.toSet());
+                .filter(n -> n.getVersion().after(Version.V_8_16_0))
```
[the remark this answers is not in the document]
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java:207]: will change 

> So the only thing we can use is release version.

to

> So the only thing we can use is release version. We added systemIndexMappingsVersion to the transport action in 8.16.1 and 8.17.0, so include nodes with version > 8.16.0.

File: qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/SystemIndexMappingUpgradeIT.java (hunk 3 above)
Comment (ldematte) [reply qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/SystemIndexMappingUpgradeIT.java:98]: Couldn't find one, but I'll look again

**Comment by ldematte** [comment]
> If this is being backported to 8.16, the backport PR should be in place too, and the backport transport version referenced in this PR

Yes, definitely, I wanted to have this in the final form before backporting.

File: server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java:207]: Its worth mentioning that this affect 8.11+, but we only fix it up in 8.16.1+

**Pull request merged**